src/motor_hardware.cc

```diff
@@ -607,6 +613,93 @@ float MotorHardware::calculateBatteryPercentage(float voltage, int cells, const
+    static uint8_t status = sensor_msgs::BatteryState::POWER_SUPPLY_STATUS_UNKNOWN;
+    static float voltage_avg; // Averaged measured voltage with EWMA
+    static ros::Time last_loop_time;
+    static uint16_t sample_num = 0;
+    static ros::Duration trickle_dur(0);
+    // Initialize power status estimation
+    if (!init) {
+        voltage_avg = battery_voltage;
+        last_loop_time = ros::Time::now();
+        init = true;
+    }
+    // Calculate delta-time
+    ros::Time t = ros::Time::now();
+    ros::Duration dt_r = t - last_loop_time;
+    float dt = dt_r.toSec();
+    last_loop_time = t;
+    // Average measured voltage with EWMA to filter out high-frequency noise
+    voltage_avg = ALPHA * battery_voltage + BETA * voltage_avg;
+    voltage_buf.push_back(voltage_avg);
+    // Wait until there are enough samples to start estimation
+    // and voltage average has settled
+    if (sample_num < AVG_SAMPLE_NUM * 2) {
+        sample_num++;
+        return status;
+    }
+    // Past-looking Voltage average derivative
+    // This is done under the assumption that the samples contain predominantly
+    // high-frequency noise of a small and fairly constant amplitude
+    float d = (voltage_avg - voltage_buf.front()) / (AVG_SAMPLE_NUM * 2 * dt);
+
+    bool is_charging = status == sensor_msgs::BatteryState::POWER_SUPPLY_STATUS_CHARGING;
+    bool is_tri_charging = status == sensor_msgs::BatteryState::POWER_SUPPLY_STATUS_FULL;
+    bool is_v_rising = d > UNSTEADY_VOLT_RATE;
+    bool is_v_dropping = d < -UNSTEADY_VOLT_RATE;
+    bool is_v_idle = std::abs(d) < STEADY_VOLT_RATE;
+    bool is_in_nom_range = voltage_avg < SLA_AGM_V_NOM;
+    bool is_in_chg_range = voltage_avg >= SLA_AGM_V_CHG;
+    bool is_in_trickle_range = voltage_avg > (SLA_AGM_V_TRI - V_TOL) && voltage_avg < (SLA_AGM_V_TRI + V_TOL);
+    bool was_unplugged = (is_charging || is_tri_charging) && is_v_dropping;
+
+    if (is_v_rising || is_in_chg_range) {
+        // Charging
+        status = sensor_msgs::BatteryState::POWER_SUPPLY_STATUS_CHARGING;
+        trickle_dur = ros::Duration(0);
+    } else if (was_unplugged || (is_v_idle && is_in_nom_range && !is_charging)) {
+        // Not charging
+        status = sensor_msgs::BatteryState::POWER_SUPPLY_STATUS_NOT_CHARGING;
+        trickle_dur = ros::Duration(0);
+    } else if (is_v_idle && is_in_trickle_range) {
+        // Trickle charging
+        if (trickle_dur > TRICKLE_DUR_THRESH) {
+            status = sensor_msgs::BatteryState::POWER_SUPPLY_STATUS_FULL;
+        } else {
+            trickle_dur += dt_r;
+        }
+    } else if (is_v_dropping && is_in_nom_range && !is_charging) {
+        // Discharging
+        status = sensor_msgs::BatteryState::POWER_SUPPLY_STATUS_DISCHARGING;
+        trickle_dur = ros::Duration(0);
+    }
+    printf("%f, %f, %f, %f, %f, %d\n", ros::Time::now().toSec(), battery_voltage, voltage_avg, d, trickle_dur.toSec(), status);
```

Comment:
We don't use raw printf except prior to start of ROS situations.  See other uses in this same file of  ROS_INFO

Reply:
Oh shoot, I forgot to remove that after debugging - will remove.

Reply:
Remove.